### ISSUE TITLE
Project Settings Crash with Feature Test - 8-1-16

### ISSUE THREAD

**Issue by RodneyBaker** [turtletooth/MotionMonsterPortable#3]
Thanks for the 8-1-16 release of OpenToonz portable.  There are some great experimental additions included and I'm very interested in your work relating to creating and sharing Projects.

I note that upon selecting File/Project Settings I get a consistent crash.
I initially thought that might relate to creating a new project with a name but this is both with and without a new Project name being initiated.

Keep up the great work.

Added:  Camera Capture is now working in 8-1-16  There are some issues with initial launch of (missing) cameras and changing cameras but those relate to code in the official OpenToonz release the portable is inheriting.  Fixes are in the works for those @shun-iwasawa said he was going to add a 'No Camera' item to the dropdown as well as have the user's selection persist so that it doesn't have to continuously be reset.

**Comment by turtletooth** [comment]
I figured it out- it is related to the multiple project roots.  It is trying to look in your desktop or documents for a folder called "OpenToonz".  The temporary fix is to make a folder there called that.  I will update the portable soon.

**Comment by RodneyBaker** [comment]
Nice sleuthing.

The underlying problem (of saving all assets with a Project is a persistent one.  I have a few recommendations... one of which might be to organize the list of options under the File Menu into a more logical order given the desired or optimal workflow that will save all resources in a way that guarantees the  project can be reopened again with all assets intact. The logical order (to me) would appear to be having New Project, Project Settings and Save Project immediately under File.  Similar options for Scenes would be next, followed by Levels and then Rendering and Preferences.  

I think I can see why the current order might be present because it is likely assumed that the Project is already identified and setup by the user and so Scene and Level are of a higher priority.  My response to that would be that if such the case the order is still important but there might be other ways to de-emphasize those options that are already in place such as making clean menu items slightly darker and/or putting an asterix next to the dirty ones that haven't been saved since something changed. I know of several programs that use the asterix as a means to signal what hasn't yet been saved.

The issue I see from time to time is that Save All doesn't appear to save drawings/levels or if it does it doesn't save them in the same general path of the Project or Scene.  That's a painful experience and I've taken to making sure I at least render once to make sure I have something to show for work accomplished.  I expect the loss of work in this fashion is is frustrating enough for some to set OpenToonz aside.   I readily admit that this loss is likely due to me forgetting to explicitly create/name a project but there should be a clear way to recover from this oversight such as a Save Project As option.

**Comment by turtletooth** [comment]
I have actually never lost work this way.  Can you give me some steps to replicate the lossed levels?  In my experience, levels have always been saved in the drawings and extras folders under a project (drawings for vector and toonz raster, and extras for raster).  

I definitely want to try and fix this but I need to replicate the issue.

**Comment by RodneyBaker** [comment]
This mostly relates to workflow and I confess that I don't know what 'the proper workflow' is.  I must assume others won't either.

Before I address my issue it might be best to examine the environment the issue lives in...

There are many separate issues/elements that relate to this one of which to me is that while Projects and Project roots appear to be very important there is no direct emphasis on creating/opening/saving Projects.  

I will not assume that upon opening OpenToonz a user will know to immediately go down into the File menu listing, locate New Project and then correctly name the Project and set the proper options.  And yet, this appears to be the ideal thing for a user to do.  Before they draw a single line... before they do anything... they must setup their Project.   This must either not be true or be an area for considerable improvement.

Suggestions for improvement in this area might be:
- Move the Project options from down in the menu to the top of the menu under the File header.  This to emphasize the importance of setting up Project and Project root for everything that will be related to the session of work the user will do.

Consideration of an Open Project option has something of an issue in that there really isn't a Project file per se.  A Project within OpenToonz appears to be more the framework of directories where the assets of the project reside relative to the root of the Project.  Do most users know that in order to open a Project they really need to open a Scene from that Project?   I'm missing something important here.

Where files appear to go missing seems mainly to be when the checkbox options in the New Project dialog are unchecked.  I must assume the levels/drawing get saved to the primary path or sandbox in the directory C:\Opentoon 1.0 Stuff that is created upon startup as that is where they can generally be found when they turn up missing.

I am going to go back through some of the video tutorials that were originally posted shortly after the release of OpenToonz to see what workflow was presented to new users of OpenToonz.  There was some confusion on this matter especially given there was no Save All option at the time and users were expected to save everything individually... level by level... scene by scene... etc.  

Assuming I may not find an example of ideal Project setup and saving I'm certainly interested in seeing that workflow.  The secondary interest being how to promote that workflow in such a way that new users easily grasp that flow.

As for steps to reproduce 'lost work'... this should be fairly straightforward but I'll have to test again myself as it all depends on how things are setup in a project and what is loaded at the outset of using OpenToonz. 

Launch OpenToonz
Start drawing (if necessary click a cell in the xsheet first)
Determine to save this work by going to the File menu
See Save Scene and select that
Save your Scene in a location you'll be able to remember (because while the scene may return the drawing level associated with it may not)
At this point we can do several things to prompt a user save... we might attempt to close our project (a dialogue box should appear letting us know that we havent saved everything and granting various options.   Another might be to consider the user will look to the File Menu and opt for Save All (because it should logically save all of their work right?).  This Save All is where the problem arises for me the most because I tend to use Save Scene initially followed by a Save All.  The issue here is that a Project name has not been established yet and so OpenToonz saves things were is instructed to save them (The Scene file was already saved to a location identified by the user so gets resaved at that same location.  The drawings/level will very likely get saved in the C:\Opentoonz 1.0 Stuff directory in the last saved Project folder or the sandbox in the drawings folder.   

It is not clear to me (and I assume to many others why there is a difference in nonmenclature between what is seen in the user interface and what is saved on disk.  We can create a 'New Project' but there is no place to 'Save Project' or 'Open Project'.   We can draw drawings and save them as a Level but there is no directory for Levels... but there is a drawings folder.  

All of this to suggest that there is a lot going on that isn't known by the user (and that can be good!) but not if they have to struggle against workflow established by the user interface that won't always appear to work.

The bottom line of this seems to be (as others have pointed out) that OpenToonz user interface is optimized for use by studios where an individual only does one aspect of the work and doesn't need to know about areas others will use.  This is in stark contrast to the individual who will be taking on all aspects of a project from start to finish and/or those that need to share files with others.

Sorry to make this long (most of my posts are long and I'm trying to shorten them).
I'll see if I can't record a few videos that outline various workflows both optimum and otherwise.

**Comment by RodneyBaker** [comment]
I should add...

The primary means I use to open a 'Project' that was previously saved is to use the 'Open Recently Saved Scene' in the File Menu.  This opens the Scene with the structure of the Project and (usually) the levels too.

I use the term 'Project a bit loosely here because it's not the project we are opening so much as the Scene which in turn invokes OpenToonz to open all associated files.  This relates to why @shun-iwasawa has stated the importance of the Scene folder being stored within the root.  Of course that relates directly to the feature we are commenting on now that allows us to move that root which  I think is  important to be able to do.

**Comment by RodneyBaker** [comment]
There is another interesting thing that is presented to the user that can occassionally result in missing files and that relates to what the user opts to do if in one project and attempting to open resources of another project.

The scenario is this:
Create new project
Draw something
Now attempt to open a previous scene (via Open Recently Saved Scene).
The user should be prompted choices to Cancel, Import into Project or Change Projects. 

Cancel is straightforward
Change projects appears to be straightfoward too but the user may not realize they are about to lose work.  (i.e. room for improvement or perhaps a secondary warning)
Import into Current Project also appears to displace current work in favor of the Scene that is imported.

 This area of interest may be a primary source of file loss.

**Comment by turtletooth** [comment]
Issue fixed.  Closing.
